lib/core/proxy.c

```diff
@@ -557,21 +575,36 @@ static int on_1xx(h2o_httpclient_t *client, int version, int status, h2o_iovec_t
     return 0;
 }
 
-static void proceed_request(h2o_httpclient_t *client, size_t written, int is_end_stream)
+static void proceed_request(h2o_httpclient_t *client, size_t written, h2o_send_state_t send_state)
 {
     struct rp_generator_t *self = client->data;
+    if (self == NULL) {
+        return;
+    }
+    if (!h2o_send_state_is_in_progress(send_state)) {
+        detach_client(self);
+    }
     if (self->src_req->proceed_req != NULL)
-        self->src_req->proceed_req(self->src_req, written, is_end_stream);
+        self->src_req->proceed_req(self->src_req, written, send_state);
 }
 
 static int write_req(void *ctx, h2o_iovec_t chunk, int is_end_stream)
 {
     struct rp_generator_t *self = ctx;
+    h2o_httpclient_t *client = self->client;
+
+    if (client == NULL) {
+        return -1;
```

Comment:
I wonder if -1 is the right return code, here: won't the proxy layer generate an error response with the right context, for example: https://github.com/h2o/h2o/pull/2010/files#diff-302e03bf5939f9d70777e2647bdb940aR624

If we return 0, the frontend layer will silently ignore the error until the appropriate response has been generated.

Reply:
I reread this comment and I'm confused.

> won't the proxy layer generate an error response with the right context, for example: 

what did you want to show me with the above link?

> If we return 0, the frontend layer will silently ignore the error until the appropriate response has been generated.

I'm not sure I understand, but to me it's more natural to return -1 and let our protocol handlers check the return value. For example, on h2 frontend, we should send RST_STREAM if sending request body failed. OTOH on h1 frontend, if we've already sent the response header we should silently ignore the error, otherwise we should send an error that means `failed to send request body` I think. Anyway could you elaborate your suggestion a bit more?

Reply:
> what did you want to show me with the above link?

My intention was to show the proxy layer sending an http response, but bear with me...

> I'm not sure I understand, but to me it's more natural to return -1 and let our protocol handlers check the return value. For example, on h2 frontend, we should send RST_STREAM if sending request body failed. OTOH on h1 frontend, if we've already sent the response header we should silently ignore the error, otherwise we should send an error that means failed to send request body I think. Anyway could you elaborate your suggestion a bit more?

 ... your response here made me realize that there are two cases:
 1. we have already forwarded headers on the frontend socket, in which case, RST_STREAM would be appropriate
 2. we have not already forwarded headers on the frontend socket, in which case, it might be better to let proxy.c generate a 502 instead -- which was my original suggestion.

What do you think?

Reply:
Following up on this, since there are two ways to generate headers, either via proxy.c or via the frontend code, it seems that we should track whether headers have been sent in the request, and ignore headers. I'll post a proposal in the streaming bodies for http1 PR.

Reply:
> 1. we have already forwarded headers on the frontend socket, in which case, RST_STREAM would be appropriate
> 2. we have not already forwarded headers on the frontend socket, in which case, it might be better to let proxy.c generate a 502 instead -- which was my original suggestion.

Yes I think this is the appropriate behavior, and the current implementation does it :)

Reply:
How does the current implementation know that headers have been forwarded already?

Reply:
I believe https://github.com/h2o/h2o/blob/24d2036a5a9a3b7e218f8c82d3e92b5639760e65/lib/http2/connection.c#L655-L659 does that, but I realized I might be wrong. please let me check my code again and wait your proposal

Reply:
That makes sense @i110 , also, as you explained offline, the http1 stack can look at `conn->_ostr_final.sent_headers`